### PR TITLE
C# conformance tests

### PR DESCRIPTION
**Pull request by jskeet**
I had this in two separate branches, but I fail at git, so I've introduced the tests and fixed them in the same PR.

In short:

- Added conformance tests
- Notice when we don't have enough data to read an embedded message
- Consume unknown fields (instead of ignoring the tag and then reading the data as another tag!)

### CODE REVIEW

File: csharp/src/Google.Protobuf.Conformance/Program.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            while (RunTest(input, output))
+            {
+                count++;
+            }
+            Console.Error.WriteLine("Received EOF after {0} tests", count);
+        }
+
+        private static bool RunTest(BinaryReader input, BinaryWriter output)
+        {
+            int? size = ReadInt32(input);
```
Comment (haberman) [inline csharp/src/Google.Protobuf.Conformance/Program.cs]: Can't use input.ReadUInt32()?

**Comment by haberman** [comment]
Looks great, thanks for doing this!

File: csharp/src/Google.Protobuf.Conformance/Program.cs (hunk 1 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf.Conformance/Program.cs]: Nope - that would throw if we reached the end of the input. If we could detect whether we were at the end of the stream, we could test for that and *then* call ReadInt32 or ReadUInt32, but we can't do so easily :(

**Comment by jskeet** [comment]
I'm very glad I did - it found two significant bugs!

**Comment by jtattermusch** [comment]
LGTM after addressing the one small comment. Thanks for implementing this and good job catching the bugs.

**Comment by jskeet** [comment]
Merging as only change since last "green" was in AssemblyInfo which wouldn't affect other builds, and the C# build is green.

**Pull request merged**